### PR TITLE
hadolint Dockerfile

### PR DESCRIPTION
**Pull request by stephengroat**
## Summary

start using [`hadolint`](https://github.com/hadolint/hadolint) on `Dockerfile`s to improve syntax and performance.

## Related issues

N/A

**Checklist**:
- [ ] add related issues
- [ ] updated docs
- [ ] updated unit tests
- [ ] updated UPGRADING.md
- [ ] ready for review

### CODE REVIEW

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -90,6 +90,9 @@ jobs:
     steps:
       - name: Checkout code
         uses: actions/checkout@v2
+      - name: lint
+        run: |
+          docker run --rm --interactive hadolint/hadolint < Dockerfile
```
Comment (desimone) [inline .github/workflows/test.yaml:95]: Neat, I had not heard of hadolint before today. Seems like a great project. 

Would it be possible to also add our release dockerfiles too?

https://github.com/pomerium/pomerium/tree/master/.github

Comment (stephengroat) [reply .github/workflows/test.yaml:95]: > Would it be possible to also add our release dockerfiles too?

sure, wanted to make sure that something like this was actually desired before adding the other `Dockerfile`s too

**Comment by travisgroth** [comment]
Hi @stephengroat - any reason not to use https://github.com/marketplace/actions/hadolint-action?

**Comment by stephengroat** [comment]
> Hi @stephengroat - any reason not to use https://github.com/marketplace/actions/hadolint-action?

because I didn't know about it :-)

switching now

**Comment by travisgroth** [comment]
Thanks @stephengroat.  LGTM.  Would you mind adding the other Dockerfiles?

**Comment by stephengroat** [comment]
sure

same `lint` step or is there a task that would be better for the `goreleaser` based stuff?

**Comment by travisgroth** [comment]
> sure
> 
> same `lint` step or is there a task that would be better for the `goreleaser` based stuff?

I think same job but they'd have to be individual steps with that action.

File: .github/Dockerfile-release.arm32v6 [hunk 2]
```diff
@@ -1,10 +1,15 @@
...
 FROM busybox:latest as build
 RUN touch /config.yaml
 
+# hadolint ignore:DL3006
 FROM arm32v7/alpine
 ENV AUTOCERT_DIR /data/autocert
 WORKDIR /pomerium
-COPY --from=multiarch/qemu-user-static /usr/bin/qemu-aarch64-static /usr/bin/
+# hadolint ignore:DL3006
+FROM multiarch/qemu-user-static as qemu-user-static
```
Comment (travisgroth) [inline .github/Dockerfile-release.arm32v6:10]: Should this be above `FROM arm32v7/alpine`?  My understanding is the current ordering would use `multiarch/qemu-user-static` is the base of the final image, which I don't think we want.

File: .github/Dockerfile-release.arm32v7 [hunk 3]
```diff
@@ -1,10 +1,15 @@
...
 FROM busybox:latest as build
 RUN touch /config.yaml
 
+# hadolint ignore=DL3006
 FROM arm32v7/alpine
 ENV AUTOCERT_DIR /data/autocert
 WORKDIR /pomerium
-COPY --from=multiarch/qemu-user-static /usr/bin/qemu-aarch64-static /usr/bin/
+# hadolint ignore=DL3006
+FROM multiarch/qemu-user-static as qemu-user-static
```
Comment (travisgroth) [inline .github/Dockerfile-release.arm32v7:10]: See other comment on ordering

File: .github/Dockerfile-release.arm64v8 [hunk 4]
```diff
@@ -1,10 +1,15 @@
...
 FROM busybox:latest as build
 RUN touch /config.yaml
 
+# hadolint ignore=DL3006
 FROM arm64v8/alpine
 ENV AUTOCERT_DIR /data/autocert
 WORKDIR /pomerium
-COPY --from=multiarch/qemu-user-static /usr/bin/qemu-aarch64-static /usr/bin/
+# hadolint ignore=DL3006
+FROM multiarch/qemu-user-static as qemu-user-static
```
Comment (travisgroth) [inline .github/Dockerfile-release.arm64v8:10]: See other comment on ordering

**Comment by desimone** [comment]
Hey @stephengroat just checking in on this PR

**Comment by stephengroat** [comment]
@desimone sorry about the radio silence

I actually found a bug in hadolint through this PR, so I'm trying to fix that. I'll draft this until it's fixed

**Comment by travisgroth** [comment]
@stephengroat I'm going to close this now.  Feel free to reopen when ready.

**Pull request closed without merging**